File: teaching/anthropic/anthropic-protocol-book/code/python/ch05/stream_handler.py
"""
Chapter 5: Stream Handler.

High-level event dispatcher that sits on top of :class:`SSEParser`
and provides business-level semantics for Anthropic streaming
responses.  Accumulates text, tracks content blocks, surfaces usage
stats, and routes errors.

Usage::

    handler = StreamHandler()
    handler.handle_event(parsed_event)
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Any


# ---------------------------------------------------------------------------
# Supporting types
# ---------------------------------------------------------------------------


class ContentBlockType(Enum):
    TEXT = auto()
    THINKING = auto()
    TOOL_USE = auto()
    REDACTED_THINKING = auto()
    UNKNOWN = auto()


@dataclass
class ContentBlock:
    """Represents a single content block being accumulated from a stream."""

    index: int
    block_type: ContentBlockType = ContentBlockType.UNKNOWN
    text: str = ""
    thinking: str = ""
    signature: str = ""
    tool_use_id: str = ""
    tool_name: str = ""
    tool_input: str = ""  # accumulated partial JSON
    finished: bool = False


@dataclass
class StreamState:
    """Snapshot of the current stream processing state."""

    message_id: str = ""
    model: str = ""
    role: str = ""
    content_blocks: list[ContentBlock] = field(default_factory=list)
    stop_reason: str | None = None
    stop_sequence: str | None = None
    input_tokens: int = 0
    output_tokens: int = 0
    finished: bool = False
    error: dict[str, Any] | None = None

    @property
    def text(self) -> str:
        """Concatenated text from all text-type content blocks."""
        return "".join(
            b.text
            for b in self.content_blocks
            if b.block_type == ContentBlockType.TEXT
        )

    @property
    def thinking_text(self) -> str:
        """Concatenated thinking from all thinking-type content blocks."""
        return "".join(
            b.thinking
            for b in self.content_blocks
            if b.block_type == ContentBlockType.THINKING
        )


# ---------------------------------------------------------------------------
# Event payload type (discriminated union via type field)
# ---------------------------------------------------------------------------

@dataclass
class _RawEvent:
    type: str
    index: int | None = None
    delta: dict[str, Any] | None = None
    content_block: dict[str, Any] | None = None
    message: dict[str, Any] | None = None
    usage: dict[str, Any] | None = None
    error: dict[str, Any] | None = None


# ---------------------------------------------------------------------------
# Stream Handler
# ---------------------------------------------------------------------------


class StreamHandler:
    """High-level handler for Anthropic streaming events.

    Translates raw SSE events into accumulated state and emits
    displayable text deltas to the caller.

    Usage::

        state = StreamState()
        handler = StreamHandler(state)

        for parsed_event in parser.feed(chunk):
            text = handler.handle_event(parsed_event)
            if text:
                print(text, end="", flush=True)
    """

    def __init__(self, state: StreamState | None = None) -> None:
        self._state = state or StreamState()

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    @property
    def state(self) -> StreamState:
        """Current accumulated stream state."""
        return self._state

    def handle_event(self, event: dict[str, Any]) -> str | None:
        """Dispatch a single parsed SSE event.

        Returns:
            A displayable text string if the event carries a text
            delta, or ``None`` for non-text events.
        """
        event_type = event.get("type", "")
        if not event_type:
            return None

        handler = _DISPATCH.get(event_type)
        if handler is not None:
            return handler(self, event)
        # Unknown event types are silently ignored per Anthropic's
        # versioning policy.
        return None

    # ------------------------------------------------------------------
    # Event handlers (one per Anthropic event type)
    # ------------------------------------------------------------------

    def _on_message_start(self, event: dict[str, Any]) -> None:
        msg = event.get("message", {})
        self._state.message_id = msg.get("id", "")
        self._state.model = msg.get("model", "")
        self._state.role = msg.get("role", "")
        usage = msg.get("usage", {})
        self._state.input_tokens = usage.get("input_tokens", 0)
        self._state.output_tokens = usage.get("output_tokens", 0)
        self._state.content_blocks = []

    def _on_content_block_start(self, event: dict[str, Any]) -> None:
        index = event.get("index", -1)
        block_data = event.get("content_block", {})
        block_type_str = block_data.get("type", "")

        block_type = _block_type_from_str(block_type_str)

        block = ContentBlock(index=index, block_type=block_type)

        if block_type == ContentBlockType.TOOL_USE:
            block.tool_use_id = block_data.get("id", "")
            block.tool_name = block_data.get("name", "")

        # Ensure content_blocks list is large enough
        while len(self._state.content_blocks) <= index:
            self._state.content_blocks.append(
                ContentBlock(index=len(self._state.content_blocks))
            )
        self._state.content_blocks[index] = block

    def _on_content_block_delta(self, event: dict[str, Any]) -> str | None:
        index = event.get("index", -1)
        delta = event.get("delta", {})
        delta_type = delta.get("type", "")

        # Ensure block exists
        while len(self._state.content_blocks) <= index:
            self._state.content_blocks.append(ContentBlock(index=index))
        block = self._state.content_blocks[index]

        if delta_type == "text_delta":
            text = delta.get("text", "")
            block.text += text
            if block.block_type == ContentBlockType.UNKNOWN:
                block.block_type = ContentBlockType.TEXT
            return text

        elif delta_type == "thinking_delta":
            thinking = delta.get("thinking", "")
            block.thinking += thinking
            if block.block_type == ContentBlockType.UNKNOWN:
                block.block_type = ContentBlockType.THINKING
            return None

        elif delta_type == "signature_delta":
            block.signature = delta.get("signature", "")
            return None

        elif delta_type == "input_json_delta":
            partial = delta.get("partial_json", "")
            block.tool_input += partial
            return None

        elif delta_type == "citations_delta":
            # Citations are accumulated but not returned as text
            return None

        # Unknown delta types are silently ignored.
        return None

    def _on_content_block_stop(self, event: dict[str, Any]) -> None:
        index = event.get("index", -1)
        while len(self._state.content_blocks) <= index:
            self._state.content_blocks.append(ContentBlock(index=index))
        self._state.content_blocks[index].finished = True

    def _on_message_delta(self, event: dict[str, Any]) -> None:
        delta = event.get("delta", {})
        self._state.stop_reason = delta.get("stop_reason")
        self._state.stop_sequence = delta.get("stop_sequence")
        usage = event.get("usage", {})
        self._state.output_tokens = usage.get("output_tokens", self._state.output_tokens)

    def _on_message_stop(self, event: dict[str, Any]) -> None:  # noqa: ARG002
        self._state.finished = True

    def _on_ping(self, event: dict[str, Any]) -> None:  # noqa: ARG002
        pass  # Transport-level keepalive -- nothing to do.

    def _on_error(self, event: dict[str, Any]) -> None:
        self._state.error = event.get("error", {})
        self._state.finished = True

    # ------------------------------------------------------------------
    # Convenience
    # ------------------------------------------------------------------

    def get_final_message(self) -> dict[str, Any]:
        """Construct a Message object from accumulated state.

        This mimics the structure returned by the non-streaming API.
        """
        content: list[dict[str, Any]] = []
        for block in self._state.content_blocks:
            if not block.finished:
                continue
            if block.block_type == ContentBlockType.TEXT:
                content.append({"type": "text", "text": block.text})
            elif block.block_type == ContentBlockType.THINKING:
                content.append(
                    {
                        "type": "thinking",
                        "thinking": block.thinking,
                        "signature": block.signature,
                    }
                )
            elif block.block_type == ContentBlockType.TOOL_USE:
                import json

                try:
                    tool_input = json.loads(block.tool_input) if block.tool_input else {}
                except json.JSONDecodeError:
                    tool_input = {}
                content.append(
                    {
                        "type": "tool_use",
                        "id": block.tool_use_id,
                        "name": block.tool_name,
                        "input": tool_input,
                    }
                )
            elif block.block_type == ContentBlockType.REDACTED_THINKING:
                content.append(
                    {
                        "type": "redacted_thinking",
                        "data": block.thinking,
                    }
                )

        return {
            "id": self._state.message_id,
            "type": "message",
            "role": self._state.role or "assistant",
            "model": self._state.model,
            "content": content,
            "stop_reason": self._state.stop_reason,
            "stop_sequence": self._state.stop_sequence,
            "usage": {
                "input_tokens": self._state.input_tokens,
                "output_tokens": self._state.output_tokens,
            },
        }

    def reset(self) -> None:
        """Reset all accumulated state (for reuse)."""
        self._state = StreamState()


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _block_type_from_str(s: str) -> ContentBlockType:
    """Map a string content block type to the enum."""
    mapping: dict[str, ContentBlockType] = {
        "text": ContentBlockType.TEXT,
        "thinking": ContentBlockType.THINKING,
        "redacted_thinking": ContentBlockType.REDACTED_THINKING,
        "tool_use": ContentBlockType.TOOL_USE,
    }
    return mapping.get(s, ContentBlockType.UNKNOWN)


# Dispatch table -- maps event types to handler methods
_DISPATCH: dict[str, Any] = {
    "message_start": StreamHandler._on_message_start,
    "content_block_start": StreamHandler._on_content_block_start,
    "content_block_delta": StreamHandler._on_content_block_delta,
    "content_block_stop": StreamHandler._on_content_block_stop,
    "message_delta": StreamHandler._on_message_delta,
    "message_stop": StreamHandler._on_message_stop,
    "ping": StreamHandler._on_ping,
    "error": StreamHandler._on_error,
}

File: teaching/anthropic/anthropic-protocol-book/code/node/ch05/StreamHandler.ts
/**
 * Chapter 5: Stream Handler (TypeScript).
 *
 * High-level event dispatcher that sits on top of `SSEParser` and
 * provides business-level semantics for Anthropic streaming responses.
 * Accumulates text, tracks content blocks, surfaces usage stats, and
 * routes errors.
 *
 * @example
 * ```ts
 * const handler = new StreamHandler();
 * for (const parsed of parser.feed(chunk)) {
 *   const text = handler.handleEvent(parsed.data);
 *   if (text) process.stdout.write(text);
 * }
 * ```
 */

// ---------------------------------------------------------------------------
// Supporting types
// ---------------------------------------------------------------------------

export const enum ContentBlockType {
  Text,
  Thinking,
  ToolUse,
  RedactedThinking,
  Unknown,
}

export interface ContentBlock {
  index: number;
  blockType: ContentBlockType;
  text: string;
  thinking: string;
  signature: string;
  toolUseId: string;
  toolName: string;
  toolInput: string;
  finished: boolean;
}

/** Snapshot of the current stream processing state. */
export interface StreamState {
  messageId: string;
  model: string;
  role: string;
  contentBlocks: ContentBlock[];
  stopReason: string | null;
  stopSequence: string | null;
  inputTokens: number;
  outputTokens: number;
  finished: boolean;
  error: Record<string, unknown> | null;
}

/** Return type for `StreamHandler.getFinalMessage()`. */
export interface FinalMessage {
  id: string;
  type: string;
  role: string;
  model: string;
  content: Record<string, unknown>[];
  stop_reason: string | null;
  stop_sequence: string | null;
  usage: {
    input_tokens: number;
    output_tokens: number;
  };
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function blockTypeFromStr(s: string): ContentBlockType {
  const mapping: Record<string, ContentBlockType> = {
    text: ContentBlockType.Text,
    thinking: ContentBlockType.Thinking,
    redacted_thinking: ContentBlockType.RedactedThinking,
    tool_use: ContentBlockType.ToolUse,
  };
  return mapping[s] ?? ContentBlockType.Unknown;
}

function makeContentBlock(index: number): ContentBlock {
  return {
    index,
    blockType: ContentBlockType.Unknown,
    text: "",
    thinking: "",
    signature: "",
    toolUseId: "",
    toolName: "",
    toolInput: "",
    finished: false,
  };
}

function ensureBlock(state: StreamState, index: number): ContentBlock {
  while (state.contentBlocks.length <= index) {
    state.contentBlocks.push(makeContentBlock(state.contentBlocks.length));
  }
  const block = state.contentBlocks[index];
  if (!block) return makeContentBlock(index);
  return block;
}

// ---------------------------------------------------------------------------
// Stream Handler
// ---------------------------------------------------------------------------

export class StreamHandler {
  private state: StreamState;

  constructor(state?: StreamState) {
    this.state = state ?? createEmptyState();
  }

  /** Current accumulated stream state. */
  getState(): Readonly<StreamState> {
    return this.state;
  }

  // ------------------------------------------------------------------
  // Public API
  // ------------------------------------------------------------------

  /**
   * Dispatch a single parsed SSE event.
   *
   * @returns A displayable text string if the event carries a text
   *          delta, or `undefined` for non-text events.
   */
  handleEvent(event: Record<string, unknown>): string | undefined {
    const eventType = (event["type"] as string) ?? "";
    if (!eventType) return undefined;

    switch (eventType) {
      case "message_start":
        return this.onMessageStart(event);
      case "content_block_start":
        return this.onContentBlockStart(event);
      case "content_block_delta":
        return this.onContentBlockDelta(event);
      case "content_block_stop":
        return this.onContentBlockStop(event);
      case "message_delta":
        return this.onMessageDelta(event);
      case "message_stop":
        return this.onMessageStop();
      case "ping":
        return this.onPing();
      case "error":
        return this.onError(event);
      default:
        // Unknown event types are silently ignored per versioning policy
        return undefined;
    }
  }

  /**
   * Construct a Message object from accumulated state.
   *
   * This mimics the structure returned by the non-streaming API.
   */
  getFinalMessage(): FinalMessage {
    const content: Record<string, unknown>[] = [];
    for (const block of this.state.contentBlocks) {
      if (!block.finished) continue;

      switch (block.blockType) {
        case ContentBlockType.Text:
          content.push({ type: "text", text: block.text });
          break;
        case ContentBlockType.Thinking:
          content.push({
            type: "thinking",
            thinking: block.thinking,
            signature: block.signature,
          });
          break;
        case ContentBlockType.ToolUse: {
          let toolInput: unknown = {};
          try {
            toolInput = block.toolInput ? JSON.parse(block.toolInput) : {};
          } catch {
            // partial JSON -- use raw string
            toolInput = block.toolInput;
          }
          content.push({
            type: "tool_use",
            id: block.toolUseId,
            name: block.toolName,
            input: toolInput,
          });
          break;
        }
        case ContentBlockType.RedactedThinking:
          content.push({
            type: "redacted_thinking",
            data: block.thinking,
          });
          break;
      }
    }

    return {
      id: this.state.messageId,
      type: "message",
      role: this.state.role || "assistant",
      model: this.state.model,
      content,
      stop_reason: this.state.stopReason,
      stop_sequence: this.state.stopSequence,
      usage: {
        input_tokens: this.state.inputTokens,
        output_tokens: this.state.outputTokens,
      },
    };
  }

  /** Concatenated text from all text-type content blocks. */
  getText(): string {
    let text = "";
    for (const b of this.state.contentBlocks) {
      if (b.blockType === ContentBlockType.Text) {
        text += b.text;
      }
    }
    return text;
  }

  /** Concatenated thinking from all thinking-type content blocks. */
  getThinkingText(): string {
    let thinking = "";
    for (const b of this.state.contentBlocks) {
      if (b.blockType === ContentBlockType.Thinking) {
        thinking += b.thinking;
      }
    }
    return thinking;
  }

  /** Reset all accumulated state (for reuse). */
  reset(): void {
    this.state = createEmptyState();
  }

  // ------------------------------------------------------------------
  // Event handlers
  // ------------------------------------------------------------------

  private onMessageStart(event: Record<string, unknown>): undefined {
    const msg = (event["message"] as Record<string, unknown>) ?? {};
    this.state.messageId = (msg["id"] as string) ?? "";
    this.state.model = (msg["model"] as string) ?? "";
    this.state.role = (msg["role"] as string) ?? "";
    const usage = (msg["usage"] as Record<string, unknown>) ?? {};
    this.state.inputTokens = (usage["input_tokens"] as number) ?? 0;
    this.state.outputTokens = (usage["output_tokens"] as number) ?? 0;
    this.state.contentBlocks = [];
    return undefined;
  }

  private onContentBlockStart(event: Record<string, unknown>): undefined {
    const index = (event["index"] as number) ?? -1;
    const blockData =
      (event["content_block"] as Record<string, unknown>) ?? {};
    const blockTypeStr = (blockData["type"] as string) ?? "";
    const blockType = blockTypeFromStr(blockTypeStr);

    const block = makeContentBlock(index);
    block.blockType = blockType;

    if (blockType === ContentBlockType.ToolUse) {
      block.toolUseId = (blockData["id"] as string) ?? "";
      block.toolName = (blockData["name"] as string) ?? "";
    }

    while (this.state.contentBlocks.length <= index) {
      this.state.contentBlocks.push(
        makeContentBlock(this.state.contentBlocks.length),
      );
    }
    this.state.contentBlocks[index] = block;
    return undefined;
  }

  private onContentBlockDelta(
    event: Record<string, unknown>,
  ): string | undefined {
    const index = (event["index"] as number) ?? -1;
    const delta = (event["delta"] as Record<string, unknown>) ?? {};
    const deltaType = (delta["type"] as string) ?? "";

    const block = ensureBlock(this.state, index);

    switch (deltaType) {
      case "text_delta": {
        const text = (delta["text"] as string) ?? "";
        block.text += text;
        if (block.blockType === ContentBlockType.Unknown) {
          block.blockType = ContentBlockType.Text;
        }
        return text;
      }
      case "thinking_delta": {
        const thinking = (delta["thinking"] as string) ?? "";
        block.thinking += thinking;
        if (block.blockType === ContentBlockType.Unknown) {
          block.blockType = ContentBlockType.Thinking;
        }
        return undefined;
      }
      case "signature_delta": {
        block.signature = (delta["signature"] as string) ?? "";
        return undefined;
      }
      case "input_json_delta": {
        const partial = (delta["partial_json"] as string) ?? "";
        block.toolInput += partial;
        return undefined;
      }
      default:
        // Unknown delta types are silently ignored
        return undefined;
    }
  }

  private onContentBlockStop(
    event: Record<string, unknown>,
  ): undefined {
    const index = (event["index"] as number) ?? -1;
    const block = ensureBlock(this.state, index);
    block.finished = true;
    return undefined;
  }

  private onMessageDelta(event: Record<string, unknown>): undefined {
    const delta = (event["delta"] as Record<string, unknown>) ?? {};
    this.state.stopReason = (delta["stop_reason"] as string) ?? null;
    this.state.stopSequence = (delta["stop_sequence"] as string) ?? null;
    const usage = (event["usage"] as Record<string, unknown>) ?? {};
    this.state.outputTokens =
      (usage["output_tokens"] as number) ?? this.state.outputTokens;
    return undefined;
  }

  private onMessageStop(): undefined {
    this.state.finished = true;
    return undefined;
  }

  private onPing(): undefined {
    // Transport-level keepalive -- nothing to do.
    return undefined;
  }

  private onError(event: Record<string, unknown>): undefined {
    this.state.error = (event["error"] as Record<string, unknown>) ?? {};
    this.state.finished = true;
    return undefined;
  }
}

// ---------------------------------------------------------------------------
// Internal factory
// ---------------------------------------------------------------------------

function createEmptyState(): StreamState {
  return {
    messageId: "",
    model: "",
    role: "",
    contentBlocks: [],
    stopReason: null,
    stopSequence: null,
    inputTokens: 0,
    outputTokens: 0,
    finished: false,
    error: null,
  };
}
